### ISSUE TITLE
Feature request: builtin array holding ($1,$2,$3,...)

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#13412]
Migrated from [rt.perl.org#120521](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=120521) (status was 'resolved')

Searchable as RT120521$

**Comment by p5pRT** [comment]
### From @epa

#### Created by @epa



It would be useful to have a builtin array variable which holds the
values captured by the last successful regexp match\.  While you can
use m// in list context to return a list of matches\, this doesn't help
so much when you want to get an array of all matches on the RHS of a
s///\.

Suppose the builtin array is called @&#8203;CAPTURE\.  Then

&nbsp;    $\_ = 'abc';
&nbsp;    /\(\.\)\(\.\)\(\.\)/ or die;
&nbsp;    say scalar @&#8203;CAPTURE; \# prints 3
&nbsp;    say $CAPTURE\[0\];     \# prints a

There is a fair bit of boilerplate code with \($1\,$2\,$3\,$4\,$5\) up to
some arbitrary limit\, which would be made simpler and correct if there
were a way to get all captures as an array\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=wishlist

Site configuration information for perl 5.16.3:

Configured by Red Hat, Inc. at Fri May  3 12:10:03 UTC 2013.

Summary of my perl5 (revision 5 version 16 subversion 3) configuration:
   
  Platform:
    osname=linux, osvers=2.6.32-358.2.1.el6.x86_64, archname=x86_64-linux-thread-multi
    uname='linux buildvm-24.phx2.fedoraproject.org 2.6.32-358.2.1.el6.x86_64 #1 smp wed feb 20 12:17:37 est 2013 x86_64 x86_64 x86_64 gnulinux '
    config_args='-des -Doptimize=-O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4  -m64 -mtune=generic -Dccdlflags=-Wl,--enable-new-dtags -Dlddlflags=-shared -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4  -m64 -mtune=generic -Wl,-z,relro  -DDEBUGGING=-g -Dversion=5.16.3 -Dmyhostname=localhost -Dperladmin=root@localhost -Dcc=gcc -Dcf_by=Red Hat, Inc. -Dprefix=/usr -Dvendorprefix=/usr -Dsiteprefix=/usr/local -Dsitelib=/usr/local/share/perl5 -Dsitearch=/usr/local/lib64/perl5 -Dprivlib=/usr/share/perl5 -Dvendorlib=/usr/share/perl5/vendor_perl -Darchlib=/usr/lib64/perl5 -Dvendorarch=/usr/lib64/perl5/vendor_perl -Darchname=x86_64-linux-thread-multi -Dlibpth=/usr/local/lib64 /lib64 /usr/lib64 -Duseshrplib -Dusethreads -Duseithreads -Dusedtrace=/usr/bin/dtrace -Duselargefiles -Dd_semctl_semun -Di_db -Ui_ndbm -Di_gdbm -Di_shadow -Di_syslog -Dman3ext=3pm -Duseperlio -Dinstallusrbinperl=n -Ubincompat5005 -Uversiononly -Dpager=/usr/bin/less -isr -Dd_gethostent_r_proto -Ud_endhostent_r_proto -Ud_sethostent_r_proto -Ud_endprotoent_r_proto -Ud_setprotoent_r_proto -Ud_endservent_r_proto -Ud_setservent_r_proto -Dscriptdir=/usr/bin -Dusesitecustomize'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=define, usemultiplicity=define
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=define, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='gcc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4 -m64 -mtune=generic',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include'
    ccversion='', gccversion='4.7.2 20121109 (Red Hat 4.7.2-8)', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='gcc', ldflags =' -fstack-protector'
    libpth=/usr/local/lib64 /lib64 /usr/lib64
    libs=-lresolv -lnsl -lgdbm -ldb -ldl -lm -lcrypt -lutil -lpthread -lc -lgdbm_compat
    perllibs=-lresolv -lnsl -ldl -lm -lcrypt -lutil -lpthread -lc
    libc=, so=so, useshrplib=true, libperl=libperl.so
    gnulibc_version='2.16'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,--enable-new-dtags -Wl,-rpath,/usr/lib64/perl5/CORE'
    cccdlflags='-fPIC', lddlflags='-shared -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4 -m64 -mtune=generic -Wl,-z,relro '

Locally applied patches:
    


@INC for perl 5.16.3:
    /home/eda/lib/perl5/
    /usr/local/lib64/perl5
    /usr/local/share/perl5
    /usr/lib64/perl5/vendor_perl
    /usr/share/perl5/vendor_perl
    /usr/lib64/perl5
    /usr/share/perl5
    .


Environment for perl 5.16.3:
    HOME=/home/eda
    LANG=en_GB.UTF-8
    LANGUAGE (unset)
    LC_COLLATE=C
    LC_CTYPE=en_GB.UTF-8
    LC_MESSAGES=en_GB.UTF-8
    LC_MONETARY=en_GB.UTF-8
    LC_NUMERIC=en_GB.UTF-8
    LC_TIME=en_GB.UTF-8
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/eda/bin:/home/eda/bin:/usr/local/bin:/usr/bin:/sbin:/usr/sbin:/sbin:/usr/sbin
    PERL5LIB=/home/eda/lib/perl5/
    PERL_BADLANG (unset)
    SHELL=/bin/bash

______________________________________________________________________
This email has been scanned by the Symantec Email Security.cloud service.
For more information please visit http://www.symanteccloud.com
______________________________________________________________________

```
</details>

**Comment by p5pRT** [comment]
### From @Hugmeir

On Tue\, Nov 12\, 2013 at 2&#8203;:26 PM\, Ed Avis \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:

> \# New Ticket Created by  "Ed Avis"
> \# Please include the string&#8203;:  \[perl \#120521\]
> \# in the subject line of all future correspondence about this issue\.
> \# \<URL&#8203;: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=120521 >
>
>
> This is a bug report for perl from eda@&#8203;waniasset\.com\,
> generated with the help of perlbug 1\.39 running under perl 5\.16\.3\.
>
>
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please describe your issue here\]
>
> It would be useful to have a builtin array variable which holds the
> values captured by the last successful regexp match\.  While you can
> use m// in list context to return a list of matches\, this doesn't help
> so much when you want to get an array of all matches on the RHS of a
> s///\.
>
> Suppose the builtin array is called @&#8203;CAPTURE\.  Then
>
>     $\_ = 'abc';
>     /\(\.\)\(\.\)\(\.\)/ or die;
>     say scalar @&#8203;CAPTURE; \# prints 3
>     say $CAPTURE\[0\];     \# prints a
>
> There is a fair bit of boilerplate code with \($1\,$2\,$3\,$4\,$5\) up to
> some arbitrary limit\, which would be made simpler and correct if there
> were a way to get all captures as an array\.
>

Eh\, this looks like a feature better suited for CPAN\. You can already
implement this in just a few lines&#8203;:

package Capture \{
&nbsp;    use Tie&#8203;::Array;
&nbsp;    our @&#8203;ISA = qw\(Tie&#8203;::StdArray\);
&nbsp;    sub FETCH \{
&nbsp;        $\{\*\{$\_\[1\]\+1\}\{SCALAR\}\}
&nbsp;    \}
&nbsp;    sub FETCHSIZE \{
&nbsp;        scalar\(@&#8203;\-\) \- 1
&nbsp;    \}
\}

tie @&#8203;CAPTURE\, "Capture";

Although a proper implementation could avoid ties\, I imagine\. Also\, rather
than @&#8203;CAPTURE\, you could take @&#8203;\{^CAPTURE\}\, and make it available everywhere\.

>

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @demerphq

On 12 November 2013 20&#8203;:11\, Brian Fraser \<fraserbn@&#8203;gmail\.com> wrote&#8203;:
> On Tue\, Nov 12\, 2013 at 2&#8203;:26 PM\, Ed Avis \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
>>
>> \# New Ticket Created by  "Ed Avis"
>> \# Please include the string&#8203;:  \[perl \#120521\]
>> \# in the subject line of all future correspondence about this issue\.
>> \# \<URL&#8203;: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=120521 >
>>
>>
>> This is a bug report for perl from eda@&#8203;waniasset\.com\,
>> generated with the help of perlbug 1\.39 running under perl 5\.16\.3\.
>>
>>
>> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
>> \[Please describe your issue here\]
>>
>> It would be useful to have a builtin array variable which holds the
>> values captured by the last successful regexp match\.  While you can
>> use m// in list context to return a list of matches\, this doesn't help
>> so much when you want to get an array of all matches on the RHS of a
>> s///\.
>>
>> Suppose the builtin array is called @&#8203;CAPTURE\.  Then
>>
>>     $\_ = 'abc';
>>     /\(\.\)\(\.\)\(\.\)/ or die;
>>     say scalar @&#8203;CAPTURE; \# prints 3
>>     say $CAPTURE\[0\];     \# prints a
>>
>> There is a fair bit of boilerplate code with \($1\,$2\,$3\,$4\,$5\) up to
>> some arbitrary limit\, which would be made simpler and correct if there
>> were a way to get all captures as an array\.
>
>
> Eh\, this looks like a feature better suited for CPAN\. You can already
> implement this in just a few lines&#8203;:
>
> package Capture \{
>     use Tie&#8203;::Array;
>     our @&#8203;ISA = qw\(Tie&#8203;::StdArray\);
>     sub FETCH \{
>         $\{\*\{$\_\[1\]\+1\}\{SCALAR\}\}

Ugh\!

>     \}
>     sub FETCHSIZE \{
>         scalar\(@&#8203;\-\) \- 1
>     \}
> \}
>
> tie @&#8203;CAPTURE\, "Capture";
>
> Although a proper implementation could avoid ties\, I imagine\. Also\, rather
> than @&#8203;CAPTURE\, you could take @&#8203;\{^CAPTURE\}\, and make it available everywhere\.

$1 is a tie itself\. So @&#8203;\{^CAPTURE\} would be a tie as well\. The
difference between a core @&#8203;CAPTURE and a non\-core one is that the
non\-core one would be a tie which accesses a second tie\.

Yves





\-\- 
perl \-Mre=debug \-e "/just|another|perl|hacker/"

**Comment by p5pRT** [comment]
### From @demerphq

On 12 November 2013 18&#8203;:26\, Ed Avis \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> \# New Ticket Created by  "Ed Avis"
> \# Please include the string&#8203;:  \[perl \#120521\]
> \# in the subject line of all future correspondence about this issue\.
> \# \<URL&#8203;: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=120521 >
>
>
> This is a bug report for perl from eda@&#8203;waniasset\.com\,
> generated with the help of perlbug 1\.39 running under perl 5\.16\.3\.
>
>
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please describe your issue here\]
>
> It would be useful to have a builtin array variable which holds the
> values captured by the last successful regexp match\.  While you can
> use m// in list context to return a list of matches\, this doesn't help
> so much when you want to get an array of all matches on the RHS of a
> s///\.
>
> Suppose the builtin array is called @&#8203;CAPTURE\.  Then
>
>     $\_ = 'abc';
>     /\(\.\)\(\.\)\(\.\)/ or die;
>     say scalar @&#8203;CAPTURE; \# prints 3
>     say $CAPTURE\[0\];     \# prints a
>
> There is a fair bit of boilerplate code with \($1\,$2\,$3\,$4\,$5\) up to
> some arbitrary limit\, which would be made simpler and correct if there
> were a way to get all captures as an array\.

But there is&#8203;:

&nbsp;  @&#8203;CAPTURE= /\(\.\)\(\.\)\(\.\)/  or die;

IOW\, in scalar context a regex match returns a list of items captured\,
which can be fed into an array\.

Why would having a magic var for this be better?

cheers\,
Yves

**Comment by p5pRT** [comment]
### From @epa

Yves Orton wrote&#8203;:

>  @&#8203;CAPTURE= /\(\.\)\(\.\)\(\.\)/  or die;

>Why would having a magic var for this be better?

I mentioned one reason in the original report&#8203;: in s/// you don't have access to this\.
You have to do s/$re/foo\($1\,$2\,$3\,$4\,$5\.\.\.\)/

Another reason is that m// in list context makes it hard to distinguish between a
match which failed\, and one which succeeded but did not have capturing groups\.

&nbsp;    $\_ = 'abc';
&nbsp;    my @&#8203;CAPTURE = /abc/;
&nbsp;    say scalar @&#8203;CAPTURE;
&nbsp;    @&#8203;CAPTURE = /\(abc\)/;
&nbsp;    say scalar @&#8203;CAPTURE;

Both print '1'\.  So\, if you have some code like

&nbsp;    if \(/$re/\) \{ do\_something\($1\, $2\, $3\, $4\, $5\, \.\.\.\) \}

there is not an obvious way to turn it into the form you suggest&#8203;:

&nbsp;    @&#8203;CAPTURE = /$re/;
&nbsp;    \# um\, did the regexp succeed or not?

I am thinking of regular expressions which are specified in a configuration table\,
or generated at run time\.  So the number of capturing groups in the regexp is
not known in advance\.

But if @&#8203;CAPTURE were available as a builtin\, you could simply say

&nbsp;    if \(/$re/\) \{ do\_something\(@&#8203;CAPTURE\) \}

\-\- 
Ed Avis \<eda@&#8203;waniasset\.com>

\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_
This email has been scanned by the Symantec Email Security\.cloud service\.
For more information please visit http&#8203;://www\.symanteccloud\.com
\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_

**Comment by p5pRT** [comment]
### From @epa

I would add that in general the reason to prefer an array @&#8203;CAPTURE over individual $1\, $2\, $3
is the same reason why we usually work with arrays rather than a long list of variables\.
Beginners to programming will sometimes write code with $one\, $two\, $three\, $four and so on\.

Similarly\, think of @&#8203;ARGV\.  We could do without that and instead have $ARG1\, $ARG2\, $ARG3 etc\.
But it is neater and cleaner to provide an array interface\.

In the case of @&#8203;CAPTURE\, the array would have to be read\-only \(so you can't shift it\, etc\) but
still you can work with it using the normal idioms for manipulating arrays and lists in Perl\.

Has nobody else here ever written code like foo\($1\, $2\, $3\, $4\, $5\)?  And never wished there were
a more simple and obvious way to express it\, that won't have an arbitrary limit on the number
of captures?

\-\- 
Ed Avis \<eda@&#8203;waniasset\.com>

\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_
This email has been scanned by the Symantec Email Security\.cloud service\.
For more information please visit http&#8203;://www\.symanteccloud\.com
\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_

**Comment by p5pRT** [comment]
### From @emazep

On 2013\-11\-15 11&#8203;:55\, Ed Avis wrote&#8203;:

> Another reason is that m// in list context makes it hard to distinguish between a
> match which failed\, and one which succeeded but did not have capturing groups\.
> 
>     $\_ = 'abc';
>     my @&#8203;CAPTURE = /abc/;
>     say scalar @&#8203;CAPTURE;
>     @&#8203;CAPTURE = /\(abc\)/;
>     say scalar @&#8203;CAPTURE;
> 
> Both print '1'\.  So\, if you have some code like
> 
>     if \(/$re/\) \{ do\_something\($1\, $2\, $3\, $4\, $5\, \.\.\.\) \}
> 
> there is not an obvious way to turn it into the form you suggest&#8203;:
> 
>     @&#8203;CAPTURE = /$re/;
>     \# um\, did the regexp succeed or not?
> 
> I am thinking of regular expressions which are specified in a configuration table\,
> or generated at run time\.  So the number of capturing groups in the regexp is
> not known in advance\.
> 
> But if @&#8203;CAPTURE were available as a builtin\, you could simply say
> 
>     if \(/$re/\) \{ do\_something\(@&#8203;CAPTURE\) \}
> 

You can currently do&#8203;:

if \( my @&#8203;CAPTURE =\(\)= /$re/ \) \{ say @&#8203;CAPTURE \}

P\.S\.

Shouldn't be @&#8203;CAPTURE spelled @&#8203;CAPTURES?

\-Emanuele

**Comment by p5pRT** [comment]
### From @demerphq

On 15 November 2013 13&#8203;:02\, Emanuele Zeppieri \<emazep@&#8203;gmail\.com> wrote&#8203;:
> On 2013\-11\-15 11&#8203;:55\, Ed Avis wrote&#8203;:
>
>> Another reason is that m// in list context makes it hard to distinguish between a
>> match which failed\, and one which succeeded but did not have capturing groups\.
>>
>>     $\_ = 'abc';
>>     my @&#8203;CAPTURE = /abc/;
>>     say scalar @&#8203;CAPTURE;
>>     @&#8203;CAPTURE = /\(abc\)/;
>>     say scalar @&#8203;CAPTURE;
>>
>> Both print '1'\.  So\, if you have some code like
>>
>>     if \(/$re/\) \{ do\_something\($1\, $2\, $3\, $4\, $5\, \.\.\.\) \}
>>
>> there is not an obvious way to turn it into the form you suggest&#8203;:
>>
>>     @&#8203;CAPTURE = /$re/;
>>     \# um\, did the regexp succeed or not?
>>
>> I am thinking of regular expressions which are specified in a configuration table\,
>> or generated at run time\.  So the number of capturing groups in the regexp is
>> not known in advance\.
>>
>> But if @&#8203;CAPTURE were available as a builtin\, you could simply say
>>
>>     if \(/$re/\) \{ do\_something\(@&#8203;CAPTURE\) \}
>>
>
> You can currently do&#8203;:
>
> if \( my @&#8203;CAPTURE =\(\)= /$re/ \) \{ say @&#8203;CAPTURE \}
>
> P\.S\.
>
> Shouldn't be @&#8203;CAPTURE spelled @&#8203;CAPTURES?


Style preference\. I tend to prefer to use singular for arrays where I
tend to access sub elements more than the whole list&#8203;:

if \($CAPTURE\[1\]\) \{ \}

reads better to me than

if \($CAPTURES\[1\]\) \{ \}

Yves



\-\- 
perl \-Mre=debug \-e "/just|another|perl|hacker/"

**Comment by p5pRT** [comment]
### From @demerphq

On 15 November 2013 11&#8203;:55\, Ed Avis \<eda@&#8203;waniasset\.com> wrote&#8203;:
> Yves Orton wrote&#8203;:
>
>>  @&#8203;CAPTURE= /\(\.\)\(\.\)\(\.\)/  or die;
>
>>Why would having a magic var for this be better?
>
> I mentioned one reason in the original report&#8203;: in s/// you don't have access to this\.
> You have to do s/$re/foo\($1\,$2\,$3\,$4\,$5\.\.\.\)/
>
> Another reason is that m// in list context makes it hard to distinguish between a
> match which failed\, and one which succeeded but did not have capturing groups\.
>
>     $\_ = 'abc';
>     my @&#8203;CAPTURE = /abc/;
>     say scalar @&#8203;CAPTURE;
>     @&#8203;CAPTURE = /\(abc\)/;
>     say scalar @&#8203;CAPTURE;
>
> Both print '1'\.  So\, if you have some code like
>
>     if \(/$re/\) \{ do\_something\($1\, $2\, $3\, $4\, $5\, \.\.\.\) \}
>
> there is not an obvious way to turn it into the form you suggest&#8203;:
>
>     @&#8203;CAPTURE = /$re/;
>     \# um\, did the regexp succeed or not?
>
> I am thinking of regular expressions which are specified in a configuration table\,
> or generated at run time\.  So the number of capturing groups in the regexp is
> not known in advance\.
>
> But if @&#8203;CAPTURE were available as a builtin\, you could simply say
>
>     if \(/$re/\) \{ do\_something\(@&#8203;CAPTURE\) \}


Color me convinced\. :\-\)

If FC doesnt beat me to it \(as he so often does\) I will add support for this\.

Let the debate begin about the name\. I wont participate except to note
that there is a bit of a lack of consistency if we have

@&#8203;\+
@&#8203;\-
%\+
%\-

and then @&#8203;CAPTURE\(S\) as well\. I leave that for other to agonize over\.
Ever since mauve I refuse to play the bike\-shed game\.

Yves

\-\- 
perl \-Mre=debug \-e "/just|another|perl|hacker/"

**Comment by p5pRT** [comment]
### From @epa

Emanuele Zeppieri wrote&#8203;:

>    if \( my @&#8203;CAPTURE =\(\)= /$re/ \) \{ say @&#8203;CAPTURE \}

I don't think that works&#8203;:

&nbsp;    $\_ = 'abc';
&nbsp;    my $re = qr/abc/;
&nbsp;    if \( my @&#8203;CAPTURE =\(\)= /$re/ \) \{ say @&#8203;CAPTURE \} else \{ say 'no' \}
&nbsp;    if \(/$re/\) \{ say 'simple match' \}

For me this prints

no
simple match

\-\- 
Ed Avis \<eda@&#8203;waniasset\.com>

\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_
This email has been scanned by the Symantec Email Security\.cloud service\.
For more information please visit http&#8203;://www\.symanteccloud\.com
\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_

**Comment by p5pRT** [comment]
### From @epa

Thank you Yves for offering to implement this\.  You can call it @&#8203;BIKESHED if you want\, and I'll still be delighted\.\.\.

\-\- 
Ed Avis \<eda@&#8203;waniasset\.com>

\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_
This email has been scanned by the Symantec Email Security\.cloud service\.
For more information please visit http&#8203;://www\.symanteccloud\.com
\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_

**Comment by p5pRT** [comment]
### From @Leont

On Fri\, Nov 15\, 2013 at 11&#8203;:19 AM\, demerphq \<demerphq@&#8203;gmail\.com> wrote&#8203;:

> > Eh\, this looks like a feature better suited for CPAN\. You can already
> > implement this in just a few lines&#8203;:
> >
> > package Capture \{
> >     use Tie&#8203;::Array;
> >     our @&#8203;ISA = qw\(Tie&#8203;::StdArray\);
> >     sub FETCH \{
> >         $\{\*\{$\_\[1\]\+1\}\{SCALAR\}\}
>
> Ugh\!
>
> >     \}
> >     sub FETCHSIZE \{
> >         scalar\(@&#8203;\-\) \- 1
> >     \}
> > \}
> >
> > tie @&#8203;CAPTURE\, "Capture";
> >
> > Although a proper implementation could avoid ties\, I imagine\. Also\,
> rather
> > than @&#8203;CAPTURE\, you could take @&#8203;\{^CAPTURE\}\, and make it available
> everywhere\.
>
> $1 is a tie itself\. So @&#8203;\{^CAPTURE\} would be a tie as well\. The
> difference between a core @&#8203;CAPTURE and a non\-core one is that the
> non\-core one would be a tie which accesses a second tie\.
>

Now if only we had array vtables\, we could have implemented this in C
without having to use ties :\-/\.

Leon

**Comment by p5pRT** [comment]
### From @demerphq

The = \(\) = is superfluous\.

Yves

On 15 November 2013 13&#8203;:08\, Ed Avis \<eda@&#8203;waniasset\.com> wrote&#8203;:
> Emanuele Zeppieri wrote&#8203;:
>
>>    if \( my @&#8203;CAPTURE =\(\)= /$re/ \) \{ say @&#8203;CAPTURE \}
>
> I don't think that works&#8203;:
>
>     $\_ = 'abc';
>     my $re = qr/abc/;
>     if \( my @&#8203;CAPTURE =\(\)= /$re/ \) \{ say @&#8203;CAPTURE \} else \{ say 'no' \}
>     if \(/$re/\) \{ say 'simple match' \}
>
> For me this prints
>
> no
> simple match
>
> \-\-
> Ed Avis \<eda@&#8203;waniasset\.com>
>
> \_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_
> This email has been scanned by the Symantec Email Security\.cloud service\.
> For more information please visit http&#8203;://www\.symanteccloud\.com
> \_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_



\-\- 
perl \-Mre=debug \-e "/just|another|perl|hacker/"

**Comment by p5pRT** [comment]
### From @hvds

Ed Avis \<eda@&#8203;waniasset\.com> wrote&#8203;:
:I would add that in general the reason to prefer an array @&#8203;CAPTURE over individual $1\, $2\, $3
:is the same reason why we usually work with arrays rather than a long list of variables\.
:Beginners to programming will sometimes write code with $one\, $two\, $three\, $four and so on\.
:
:Similarly\, think of @&#8203;ARGV\.  We could do without that and instead have $ARG1\, $ARG2\, $ARG3 etc\.
:But it is neater and cleaner to provide an array interface\.
:
:In the case of @&#8203;CAPTURE\, the array would have to be read\-only \(so you can't shift it\, etc\) but
:still you can work with it using the normal idioms for manipulating arrays and lists in Perl\.
:
:Has nobody else here ever written code like foo\($1\, $2\, $3\, $4\, $5\)?  And never wished there were
:a more simple and obvious way to express it\, that won't have an arbitrary limit on the number
:of captures?

I haven't\, but I have frequently written code that makes the relevant array
using @&#8203;\+ and @&#8203;\-; I've always found that perfectly adequate so haven't felt
a need for yet another magic regexp variable\.

I'd rather see this prototyped on CPAN first\, if only because IMO we have
a definite habit of failing to get the semantics tied down in ideal form
first time round whenever we add something new\.

But I also feel that while the magic variable approach maybe made sense
when there were hardly any\, we've moved incrementally far beyond sanity&#8203;:
I would rather see us moving in the direction of a capture object that
you can ask all this sort of thing of via method calls\, like&#8203;:

&nbsp;  \# assuming a new //mauve option that means "give me an object"
&nbsp;  if \($match = /\(r\)\(e\)/mauve\) \{
&nbsp;    print "Saw \<$\_>" for $match\->captures;
&nbsp;  \}

\(Sorry I haven't really been following this discussion\, I'm in the middle
of moving house\, and will shortly be losing internet access for a few days\.\)

Hugo

**Comment by p5pRT** [comment]
### From @epa

I agree that a 'match object' might be a cleaner way to access properties of a regexp match\.
But in this case\, I don't think there is any cloudiness or room for debate about the semantics&#8203;:
$CAPTURE\[0\] is exactly equivalent to $1\, $CAPTURE\[1\] to $2\, and so on\.  The semantics of
$1\, $2\, etc are already well defined \(even if less than ideal in my view\)\, it's just a case of
providing a more convenient syntax to access them\.

\-\- 
Ed Avis \<eda@&#8203;waniasset\.com>

\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_
This email has been scanned by the Symantec Email Security\.cloud service\.
For more information please visit http&#8203;://www\.symanteccloud\.com
\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_

**Comment by p5pRT** [comment]
### From @demerphq

On 15 November 2013 15&#8203;:19\, Ed Avis \<eda@&#8203;waniasset\.com> wrote&#8203;:
> I agree that a 'match object' might be a cleaner way to access properties of a regexp match\.
> But in this case\, I don't think there is any cloudiness or room for debate about the semantics&#8203;:
> $CAPTURE\[0\] is exactly equivalent to $1\, $CAPTURE\[1\] to $2\, and so on\.  The semantics of
> $1\, $2\, etc are already well defined \(even if less than ideal in my view\)\, it's just a case of
> providing a more convenient syntax to access them\.

I agree\. We introduced %\- and %\+ without difficulty\. I agree with hv
in general with regard to the regex engine\. However I think this case
is different\.

\-\- 
perl \-Mre=debug \-e "/just|another|perl|hacker/"

**Comment by p5pRT** [comment]
### From @emazep

\[Sorry for the private mail Ed\, it was intended for the list\]

On 2013\-11\-15 13&#8203;:08\, Ed Avis wrote&#8203;:

> Emanuele Zeppieri wrote&#8203;:
>
>>    if \( my @&#8203;CAPTURE =\(\)= /$re/ \) \{ say @&#8203;CAPTURE \}
>
> I don't think that works&#8203;:

Indeed you are right\, sorry again :\-\)

However\, give me a second attempt&#8203;:

$\_ = 'abc';

sub test \{
&nbsp;    if \( my @&#8203;CAPTURE = /$\_\[0\]/ \) \{ f\( defined $1 ? @&#8203;CAPTURE : \(\) \) \}
&nbsp;    else \{ say 'NO' \}
\}

sub f \{ say join '\-'\, @&#8203;\_ \}

test qr/abc/;       \# match w/o captures&#8203;: f is called with \(\)
test qr/\(a\)\(b\)\(c\)/; \# match w/ captures : f is called with qw\(a b c\)
test qr/abcz/;      \# it does not match : f is not called

What's the missing case?

**Comment by p5pRT** [comment]
### From ema.zep@libero.it

On 2013\-11\-15 13&#8203;:02\, Emanuele Zeppieri wrote&#8203;:
> On 2013\-11\-15 11&#8203;:55\, Ed Avis wrote&#8203;:
> 
>> Another reason is that m// in list context makes it hard to distinguish between a
>> match which failed\, and one which succeeded but did not have capturing groups\.
>>
>>     $\_ = 'abc';
>>     my @&#8203;CAPTURE = /abc/;
>>     say scalar @&#8203;CAPTURE;
>>     @&#8203;CAPTURE = /\(abc\)/;
>>     say scalar @&#8203;CAPTURE;
>>
>> Both print '1'\.  So\, if you have some code like
>>
>>     if \(/$re/\) \{ do\_something\($1\, $2\, $3\, $4\, $5\, \.\.\.\) \}
>>
>> there is not an obvious way to turn it into the form you suggest&#8203;:
>>
>>     @&#8203;CAPTURE = /$re/;
>>     \# um\, did the regexp succeed or not?
>>
>> I am thinking of regular expressions which are specified in a configuration table\,
>> or generated at run time\.  So the number of capturing groups in the regexp is
>> not known in advance\.
>>
>> But if @&#8203;CAPTURE were available as a builtin\, you could simply say
>>
>>     if \(/$re/\) \{ do\_something\(@&#8203;CAPTURE\) \}
>>
> 
> You can currently do&#8203;:
> 
> if \( my @&#8203;CAPTURE =\(\)= /$re/ \) \{ say @&#8203;CAPTURE \}

No\, it doesn't work\, sorry\!

**Comment by p5pRT** [comment]
### From @emazep

On 2013\-11\-15 13&#8203;:08\, Ed Avis wrote&#8203;:

> Emanuele Zeppieri wrote&#8203;:
>
>>    if \( my @&#8203;CAPTURE =\(\)= /$re/ \) \{ say @&#8203;CAPTURE \}
>
> I don't think that works&#8203;:

Indeed you are right\, sorry again :\-\)

However\, give me a second attempt&#8203;:

$\_ = 'abc';

sub test \{
&nbsp;    if \( my @&#8203;CAPTURE = /$\_\[0\]/ \) \{ f\( defined $1 ? @&#8203;CAPTURE : \(\) \) \}
&nbsp;    else \{ say 'NO' \}
\}

sub f \{ say join '\-'\, @&#8203;\_ \}

test qr/abc/;       \# match w/o captures&#8203;: f is called with \(\)
test qr/\(a\)\(b\)\(c\)/; \# match w/ captures : f is called with qw\(a b c\)
test qr/abcz/;      \# it does not match : f is not called

What's the missing case?

**Comment by p5pRT** [comment]
### From @emazep

On 2013\-11\-15 13&#8203;:08\, Ed Avis wrote&#8203;:

> Emanuele Zeppieri wrote&#8203;:
>
>>    if \( my @&#8203;CAPTURE =\(\)= /$re/ \) \{ say @&#8203;CAPTURE \}
>
> I don't think that works&#8203;:

Indeed you are right\, sorry again :\-\)

However\, give me a second attempt&#8203;:

$\_ = 'abc';

sub test \{
&nbsp;    if \( my @&#8203;CAPTURE = /$\_\[0\]/ \) \{ f\( defined $1 ? @&#8203;CAPTURE : \(\) \) \}
&nbsp;    else \{ say 'NO' \}
\}

sub f \{ say join '\-'\, @&#8203;\_ \}

test qr/abc/;       \# match w/o captures&#8203;: f is called with \(\)
test qr/\(a\)\(b\)\(c\)/; \# match w/ captures : f is called with qw\(a b c\)
test qr/abcz/;      \# it does not match : f is not called

What's the missing case?

**Comment by p5pRT** [comment]
### From cm.perl@abtela.com

Le 15/11/2013 11&#8203;:55\, Ed Avis a écrit :
> Yves Orton wrote&#8203;:
>
>>   @&#8203;CAPTURE= /\(\.\)\(\.\)\(\.\)/  or die;
>
>> Why would having a magic var for this be better?
>
> I mentioned one reason in the original report&#8203;: in s/// you don't have access to this\.
> You have to do s/$re/foo\($1\,$2\,$3\,$4\,$5\.\.\.\)/

this is a problem only it you don't know beforehand the number of 
capturing groups\. Could you give an example ?

>
> Another reason is that m// in list context makes it hard to distinguish between a
> match which failed\, and one which succeeded but did not have capturing groups\.
>
>      $\_ = 'abc';
>      my @&#8203;CAPTURE = /abc/;
>      say scalar @&#8203;CAPTURE;
>      @&#8203;CAPTURE = /\(abc\)/;
>      say scalar @&#8203;CAPTURE;
>
> Both print '1'\.  So\, if you have some code like
>
>      if \(/$re/\) \{ do\_something\($1\, $2\, $3\, $4\, $5\, \.\.\.\) \}
>
> there is not an obvious way to turn it into the form you suggest&#8203;:
>
>      @&#8203;CAPTURE = /$re/;
>      \# um\, did the regexp succeed or not?

if \(my @&#8203;CAPTURE = /$re/\) \{
&nbsp;   \.\.\.
\}

>
> I am thinking of regular expressions which are specified in a configuration table\,
> or generated at run time\.  So the number of capturing groups in the regexp is
> not known in advance\.

Again\, can you provide an example ? I generally use either named capture 
groups\, as in

my \($header\, $param1\, $param2\);
if \(m/\.\.\.\(?\<header>\.\.\.\)\.\.\.\(?\<param1>\)\.\.\.\(?\<param2>\)/\) \{
&nbsp;    do\_someting\_with\_header\($header\);
&nbsp;    process\_params\($param1\, $param2\);
\}

or named targets\, as in

if \(my \($header\, $param1\, $param2\) = m/\.\.\.\./\) \{
&nbsp;    do\_someting\_with\_header\($header\);
&nbsp;    process\_params\($param1\, $param2\);
\}

@&#8203;CAPTURE as proposed here strikes me as too rigid for me in most cases 
but I might be missing something\.\.\.

>
> But if @&#8203;CAPTURE were available as a builtin\, you could simply say
>
>      if \(/$re/\) \{ do\_something\(@&#8203;CAPTURE\) \}
>

\-\-Christian

**Comment by p5pRT** [comment]
### From cm.perl@abtela.com

Le 15/11/2013 11&#8203;:55\, Ed Avis a écrit :
> Another reason is that m// in list context makes it hard to distinguish between a
> match which failed\, and one which succeeded but did not have capturing groups\.
>
>      $\_ = 'abc';
>      my @&#8203;CAPTURE = /abc/;
>      say scalar @&#8203;CAPTURE;
>      @&#8203;CAPTURE = /\(abc\)/;
>      say scalar @&#8203;CAPTURE;
>
> Both print '1'\.  So\, if you have some code like

You can always add a capturing group\, for instance an empty lookahead :

Taisha&#8203;:\~/tttmp $ perl \-E 'for \(@&#8203;ARGV\) \{ print; if \(my @&#8203;CAPTURE = 
/123\(?=\)/\) \{ print qq\{ match\\n\} \} else \{ print qq\{ fail\\n\}\}\}'  123 abc a123
123 match
abc fail
a123 match
Taisha&#8203;:\~/tttmp $

\-\-Christian

**Comment by p5pRT** [comment]
### From @emazep

On Fri\, Nov 15\, 2013 at 1&#8203;:04 PM\, demerphq \<demerphq@&#8203;gmail\.com> wrote&#8203;:

> On 15 November 2013 13&#8203;:02\, Emanuele Zeppieri \<emazep@&#8203;gmail\.com> wrote&#8203;:
> > On 2013\-11\-15 11&#8203;:55\, Ed Avis wrote&#8203;:
> >
> >> Another reason is that m// in list context makes it hard to distinguish
> between a
> >> match which failed\, and one which succeeded but did not have capturing
> groups\.
> >>
> >>     $\_ = 'abc';
> >>     my @&#8203;CAPTURE = /abc/;
> >>     say scalar @&#8203;CAPTURE;
> >>     @&#8203;CAPTURE = /\(abc\)/;
> >>     say scalar @&#8203;CAPTURE;
> >>
> >> Both print '1'\.  So\, if you have some code like
> >>
> >>     if \(/$re/\) \{ do\_something\($1\, $2\, $3\, $4\, $5\, \.\.\.\) \}
> >>
> >> there is not an obvious way to turn it into the form you suggest&#8203;:
> >>
> >>     @&#8203;CAPTURE = /$re/;
> >>     \# um\, did the regexp succeed or not?
> >>
> >> I am thinking of regular expressions which are specified in a
> configuration table\,
> >> or generated at run time\.  So the number of capturing groups in the
> regexp is
> >> not known in advance\.
> >>
> >> But if @&#8203;CAPTURE were available as a builtin\, you could simply say
> >>
> >>     if \(/$re/\) \{ do\_something\(@&#8203;CAPTURE\) \}
> >>
> >
> > You can currently do&#8203;:
> >
> > if \( my @&#8203;CAPTURE =\(\)= /$re/ \) \{ say @&#8203;CAPTURE \}
> >
> > P\.S\.
> >
> > Shouldn't be @&#8203;CAPTURE spelled @&#8203;CAPTURES?
>
>
> Style preference\. I tend to prefer to use singular for arrays where I
> tend to access sub elements more than the whole list&#8203;:
>
> if \($CAPTURE\[1\]\) \{ \}
>
> reads better to me than
>
> if \($CAPTURES\[1\]\) \{ \}
>
> Yves
>

On a second thought\, I see at least three points where you are right&#8203;:

1\. it's a question of style preference;
2\. the plural would be inconsistent with the other arrays special
variables&#8203;: @&#8203;ARG\, @&#8203;LAST\_MATCH\_END\, etc\. \(even if you didn't say it\);
3\. it's bikeshedding galore :\-\)

**Comment by p5pRT** [comment]
### From @emazep

On 2013\-11\-15 15&#8203;:27\, Ed Avis wrote&#8203;:

>>    if \( my @&#8203;CAPTURE = /$\_\[0\]/ \) \{ f\( defined $1 ? @&#8203;CAPTURE : \(\) \) \}
>
> Yes\, that works\, or perhaps
>
>    if \(my @&#8203;CAPTURE = /$re/\) \{
>        @&#8203;CAPTURE = \(\) if not defined $1;
>        do\_something\(@&#8203;CAPTURE\);
>    \}
>
> This does seem cumbersome\, though\, and involves looking at both $1 and
the list of captures returned by the match\.
>
> At the moment my code is first doing the match\, then performing the
regexp again to get the captures;
> your version is certainly an improvement on that\!
>

Well\, here is an alternative solution\, which permits to save the $1 check
and just test the match in the most natural way\.

The trick is to wrap the given regex in extra brackets&#8203;:

if \( my @&#8203;CAPTURE = /\($re\)/ \) \{ do\_something\(@&#8203;CAPTURE\) \}

and then shift away the artificial capture in the called subroutine&#8203;:

sub do\_something \{
&nbsp;   shift;
&nbsp;   \.\.\.
\}

\(or before the sub call\)\.

Just to realize how much simpler it would be if your proposal were
available\, so in the end I cannot agree more with it\.

**Comment by p5pRT** [comment]
### From @emazep

On 2013\-11\-16 00&#8203;:21\, Emanuele Zeppieri wrote&#8203;:
> On 2013\-11\-15 15&#8203;:27\, Ed Avis wrote&#8203;:
> 
>>>    if \( my @&#8203;CAPTURE = /$\_\[0\]/ \) \{ f\( defined $1 ? @&#8203;CAPTURE : \(\) \) \}
>>
>> Yes\, that works\, or perhaps
>>
>>    if \(my @&#8203;CAPTURE = /$re/\) \{
>>        @&#8203;CAPTURE = \(\) if not defined $1;
>>        do\_something\(@&#8203;CAPTURE\);
>>    \}
>>
>> This does seem cumbersome\, though\, and involves looking at both $1 and
> the list of captures returned by the match\.
>>
>> At the moment my code is first doing the match\, then performing the
> regexp again to get the captures;
>> your version is certainly an improvement on that\!
>>
> 
> Well\, here is an alternative solution\, which permits to save the $1
> check and just test the match in the most natural way\.
> 
> The trick is to wrap the given regex in extra brackets&#8203;:
> 
> if \( my @&#8203;CAPTURE = /\($re\)/ \) \{ do\_something\(@&#8203;CAPTURE\) \}
> 
> and then shift away the artificial capture in the called subroutine&#8203;:
> 
> sub do\_something \{
>    shift;
>    \.\.\.
> \}

And this further refinement avoids the shift too&#8203;:

if \( \(undef\, my @&#8203;CAPTURE\) = /\($\_\[0\]\)/ \) \{ do\_something\(@&#8203;CAPTURE\) \}

> Just to realize how much simpler it would be if your proposal were available\, so in the end I cannot agree more with it\.

No less ugly however \(and potentially inefficient too\)\, so the above is
fully confirmed anyway\.

**Comment by p5pRT** [comment]
### From @emazep

On Sat\, Nov 16\, 2013 at 1&#8203;:10 AM\, Emanuele Zeppieri \<emazep@&#8203;gmail\.com> wrote&#8203;:
>
> On 2013\-11\-16 00&#8203;:21\, Emanuele Zeppieri wrote&#8203;:
> > On 2013\-11\-15 15&#8203;:27\, Ed Avis wrote&#8203;:
> >
> >>>    if \( my @&#8203;CAPTURE = /$\_\[0\]/ \) \{ f\( defined $1 ? @&#8203;CAPTURE : \(\) \) \}
> >>
> >> Yes\, that works\, or perhaps
> >>
> >>    if \(my @&#8203;CAPTURE = /$re/\) \{
> >>        @&#8203;CAPTURE = \(\) if not defined $1;
> >>        do\_something\(@&#8203;CAPTURE\);
> >>    \}
> >>
> >> This does seem cumbersome\, though\, and involves looking at both $1 and
> > the list of captures returned by the match\.
> >>
> >> At the moment my code is first doing the match\, then performing the
> > regexp again to get the captures;
> >> your version is certainly an improvement on that\!
> >>
> >
> > Well\, here is an alternative solution\, which permits to save the $1
> > check and just test the match in the most natural way\.
> >
> > The trick is to wrap the given regex in extra brackets&#8203;:
> >
> > if \( my @&#8203;CAPTURE = /\($re\)/ \) \{ do\_something\(@&#8203;CAPTURE\) \}
> >
> > and then shift away the artificial capture in the called subroutine&#8203;:
> >
> > sub do\_something \{
> >    shift;
> >    \.\.\.
> > \}
>
> And this further refinement avoids the shift too&#8203;:
>
> if \( \(undef\, my @&#8203;CAPTURE\) = /\($\_\[0\]\)/ \) \{ do\_something\(@&#8203;CAPTURE\) \}
>
> > Just to realize how much simpler it would be if your proposal were available\, so in the end I cannot agree more with it\.
>
> No less ugly however \(and potentially inefficient too\)\, so the above is
> fully confirmed anyway\.

Well\, this avoids the potential inefficiency at least :\-\)

if \( \(undef\, my @&#8203;CAPTURE\) = /\(\)$\_\[0\]/ \) \{ do\_something\(@&#8203;CAPTURE\) \}

**Comment by p5pRT** [comment]
### From @tamias

On Sat\, Nov 16\, 2013 at 12&#8203;:21&#8203;:18AM \+0100\, Emanuele Zeppieri wrote&#8203;:
> 
> >>    if \( my @&#8203;CAPTURE = /$\_\[0\]/ \) \{ f\( defined $1 ? @&#8203;CAPTURE : \(\) \) \}
> >
> > Yes\, that works\, or perhaps
> >
> >    if \(my @&#8203;CAPTURE = /$re/\) \{
> >        @&#8203;CAPTURE = \(\) if not defined $1;
> >        do\_something\(@&#8203;CAPTURE\);
> >    \}
> >
> > This does seem cumbersome\, though\, and involves looking at both $1 and
> > the list of captures returned by the match\.
> >

You should be aware that you cannot depend on any particular capture being
defined\, even if the match is successful&#8203;:

$\_ = "b";
my $re = qr/\(a\)|\(b\)/;

if \(my @&#8203;CAPTURE = /$re/\) \{
&nbsp;  foreach my $i \(0 \.\. $\#CAPTURE\) \{
&nbsp;    print '$'\, $i \+ 1\, "&#8203;: "\,
&nbsp;          defined $CAPTURE\[$i\] ? "" : "not "\, "defined\\n";
&nbsp;  \}
\}

Ronald

**Comment by p5pRT** [comment]
### From @emazep

On Sat\, Nov 16\, 2013 at 6&#8203;:46 AM\, Ronald J Kimball \<rjk@&#8203;tamias\.net> wrote&#8203;:
> On Sat\, Nov 16\, 2013 at 12&#8203;:21&#8203;:18AM \+0100\, Emanuele Zeppieri wrote&#8203;:
>>
>> >>    if \( my @&#8203;CAPTURE = /$\_\[0\]/ \) \{ f\( defined $1 ? @&#8203;CAPTURE : \(\) \) \}
>> >
>> > Yes\, that works\, or perhaps
>> >
>> >    if \(my @&#8203;CAPTURE = /$re/\) \{
>> >        @&#8203;CAPTURE = \(\) if not defined $1;
>> >        do\_something\(@&#8203;CAPTURE\);
>> >    \}
>> >
>> > This does seem cumbersome\, though\, and involves looking at both $1 and
>> > the list of captures returned by the match\.
>> >
>
> You should be aware that you cannot depend on any particular capture being
> defined\, even if the match is successful&#8203;:
>
> $\_ = "b";
> my $re = qr/\(a\)|\(b\)/;
>
> if \(my @&#8203;CAPTURE = /$re/\) \{
>   foreach my $i \(0 \.\. $\#CAPTURE\) \{
>     print '$'\, $i \+ 1\, "&#8203;: "\,
>           defined $CAPTURE\[$i\] ? "" : "not "\, "defined\\n";
>   \}
> \}
>
> Ronald

The error was mine of course\.

However it's just a matter of finding a reliable indicator that a
capture has happened\, and the idea still stands&#8203;:

if \( my @&#8203;CAPTURE = /$re/ \) \{
&nbsp;    do\_something\( $\#\- ? @&#8203;CAPTURE : \(\) \)
\}

Does $\#\- > 0 serve the scope or are there exceptions?

**Comment by p5pRT** [comment]
### From @druud62

On 2013\-11\-15 11&#8203;:55\, Ed Avis wrote&#8203;:
> Yves Orton wrote&#8203;:

>>   @&#8203;CAPTURE= /\(\.\)\(\.\)\(\.\)/  or die;
>
>> Why would having a magic var for this be better?
>
> I mentioned one reason in the original report&#8203;: in s/// you don't have access to this\.
> You have to do s/$re/foo\($1\,$2\,$3\,$4\,$5\.\.\.\)/

&nbsp;   @&#8203;CAPTURE = m/$re/ and s//foo\(@&#8203;CAPTURE\)/e

But what if the g\-modifier is used\, AoA?

perl \-Mstrict \-MData&#8203;::Dumper \-wle '
&nbsp;   $\_ = q\{a bc def\};
&nbsp;   my $re = q\{\(\\w?\)\(\\w?\)\(\\w?\)\};
&nbsp;   print "\[@&#8203;\-\]\\t\[@&#8203;\+\]\\t\[$1\,$2\,$3\]" while /$re/g;
'
\[0 0 1 1\]    \[1 1 1 1\]    \[a\,\,\]
\[1 1 1 1\]    \[1 1 1 1\]    \[\,\,\]
\[2 2 3 4\]    \[4 3 4 4\]    \[b\,c\,\]
\[4 4 4 4\]    \[4 4 4 4\]    \[\,\,\]
\[5 5 6 7\]    \[8 6 7 8\]    \[d\,e\,f\]
\[8 8 8 8\]    \[8 8 8 8\]    \[\,\,\]

\-\- 
Ruud

**Comment by p5pRT** [comment]
### From @mauke

On 12\.11\.2013 18&#8203;:26\, Ed Avis \(via RT\) wrote&#8203;:
> It would be useful to have a builtin array variable which holds the
> values captured by the last successful regexp match\.  While you can
> use m// in list context to return a list of matches\, this doesn't help
> so much when you want to get an array of all matches on the RHS of a
> s///\.
> 
> Suppose the builtin array is called @&#8203;CAPTURE\.  Then
> 
>     $\_ = 'abc';
>     /\(\.\)\(\.\)\(\.\)/ or die;
>     say scalar @&#8203;CAPTURE; \# prints 3
>     say $CAPTURE\[0\];     \# prints a
> 
> There is a fair bit of boilerplate code with \($1\,$2\,$3\,$4\,$5\) up to
> some arbitrary limit\, which would be made simpler and correct if there
> were a way to get all captures as an array\.

Just FYI\, that functionality is available as submatches\(\) in Data&#8203;::Munge\.


\-\- 
Lukas Mai \<plokinom@&#8203;gmail\.com>

**Comment by p5pRT** [comment]
### From @rjbs

\* hv@&#8203;crypt\.org \[2013\-11\-15T08&#8203;:44&#8203;:17\]
> But I also feel that while the magic variable approach maybe made sense
> when there were hardly any\, we've moved incrementally far beyond sanity&#8203;:
> I would rather see us moving in the direction of a capture object that
> you can ask all this sort of thing of via method calls\, like&#8203;:

I agree\.  I threw this lousy code together a few months ago and it made me
think there's hope for such an approach&#8203;:

&nbsp;  use strict;
&nbsp;  use warnings;
&nbsp;  package Regexp&#8203;::Match;

&nbsp;  sub matches \{
&nbsp;    my \($class\, undef\, $qr\, $p\) = @&#8203;\_;
&nbsp;    my $str = "$\_\[1\]"; \# stringify only once

&nbsp;    return unless $str =~ $qr;
&nbsp;    my $result = \{ p => \!\! $p \};
&nbsp;    $result\->\{str\} = $str if $p;
&nbsp;    $result\->\{match\_range\} = \[ $\-\[0\]\, $\+\[0\] \];
&nbsp;    $result\->\{g\_starts\} = \[ @&#8203;\-\[ 1 \.\. $\#\- \] \];
&nbsp;    $result\->\{g\_ends\}   = \[ @&#8203;\+\[ 1 \.\. $\#\+ \] \];
&nbsp;    for \(1 \.\. $\#\-\) \{
&nbsp;      no strict 'refs';
&nbsp;      push @&#8203;\{ $result\->\{captures\} \}\, $\{ $\_ \};
&nbsp;    \}

&nbsp;    $result\->\{capture\_hash\} = \{ %\- \};

&nbsp;    bless $result\, $class;
&nbsp;  \}

&nbsp;  sub capture \{
&nbsp;    my \($self\, $n\) = @&#8203;\_;
&nbsp;    return $\_\[0\]\{captures\}\[$n \- 1\] if $n =~ /\\A\[0\-9\]\+\\z/;
&nbsp;    return\( \($\_\[0\]\{capture\_hash\}\{$n\} // \[\]\)\->\[0\] \);
&nbsp;  \}

&nbsp;  sub capture\_list \{
&nbsp;    my \($self\, $n\) = @&#8203;\_;
&nbsp;    \# this is a mess in scalar context
&nbsp;    return $\_\[0\]\{captures\}\[$n \- 1\] if $n =~ /\\A\[0\-9\]\+\\z/;
&nbsp;    return @&#8203;\{ $\_\[0\]\{capture\_hash\}\{$n\} || \[\] \};
&nbsp;  \}

&nbsp;  sub captures \{
&nbsp;    return @&#8203;\{ $\_\[0\]\{captures\} \}\,
&nbsp;  \}

&nbsp;  sub pre\_match \{
&nbsp;    return unless $\_\[0\]\{p\};
&nbsp;    return substr $\_\[0\]\{str\}\, 0\, $\_\[0\]\{match\_range\}\[0\];
&nbsp;  \}

&nbsp;  sub match \{
&nbsp;    return unless $\_\[0\]\{p\};
&nbsp;    return substr $\_\[0\]\{str\}\, $\_\[0\]\{match\_range\}\[0\]\,
&nbsp;      $\_\[0\]\{match\_range\}\[1\] \- $\_\[0\]\{match\_range\}\[0\];
&nbsp;  \}

&nbsp;  sub post\_match \{
&nbsp;    return unless $\_\[0\]\{p\};
&nbsp;    return substr $\_\[0\]\{str\}\, $\_\[0\]\{match\_range\}\[1\];
&nbsp;  \}

&nbsp;  1;

It would also be less surprising for programmers coming from just about
/anywhere/ else to be able to say qr\(\.\.\.\)\->match\($str\)

\-\- 
rjbs

**Comment by p5pRT** [comment]
### From perl5-porters@perl.org

Ricardo Signes wrote&#8203;:
> It would also be less surprising for programmers coming from just about
> /anywhere/ else to be able to say qr\(\.\.\.\)\->match\($str\)

Why match?  Why not just qr\(\.\.\.\)\->\($str\)?

\(Then we can forget smart\-match altogether\, since \->\(\.\.\.\) is our match
operator\.\)

**Comment by p5pRT** [comment]
### From @kentfredric

On 17 November 2013 09&#8203;:53\, Father Chrysostomos \<sprout@&#8203;cpan\.org> wrote&#8203;:
> Why match?  Why not just qr\(\.\.\.\)\->\($str\)?
>
> \(Then we can forget smart\-match altogether\, since \->\(\.\.\.\) is our match
> operator\.\)


Maybe because it lends itself to later enhancements\, ie&#8203;:

qr\(\.\.\.\)\->replace\(sub\{

\}\);

May be nicer  s\{\.\.\.\}\{ code \}e;

\( Especially seeing this syntax is more recognised by code
highlighters\, some treat both sides of s\{\}\{\} as strings \)



\-\- 
Kent

**Comment by p5pRT** [comment]
### From @tamias

On Fri\, Nov 15\, 2013 at 02&#8203;:19&#8203;:38PM \+0000\, Ed Avis wrote&#8203;:
> I agree that a 'match object' might be a cleaner way to access properties of a regexp match\.
> But in this case\, I don't think there is any cloudiness or room for debate about the semantics&#8203;:
> $CAPTURE\[0\] is exactly equivalent to $1\, $CAPTURE\[1\] to $2\, and so on\.  The semantics of
> $1\, $2\, etc are already well defined \(even if less than ideal in my view\)\, it's just a case of
> providing a more convenient syntax to access them\.

$\-\[0\] and $\+\[0\] are the offsets into the string of the entire match; $\-\[1\]
and $\+\[1\] are the offsets into the string of $1\.  One could argue that
$CAPTURE\[0\] should be the entire string that was matched\, and $CAPTURE\[1\]
should be $1\.

Ronald

**Comment by p5pRT** [comment]
### From @epa

>One could argue that $CAPTURE\[0\] should be the entire string that was matched\, and $CAPTURE\[1\] should be $1\.

I'd be fine with that\.  It does remove the off\-by\-one oddity compared to $1\, etc\.

\-\- 
Ed Avis \<eda@&#8203;waniasset\.com>

\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_
This email has been scanned by the Symantec Email Security\.cloud service\.
For more information please visit http&#8203;://www\.symanteccloud\.com
\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_

**Comment by p5pRT** [comment]
### From @emazep

On Sat\, Nov 16\, 2013 at 12&#8203;:01 PM\, Emanuele Zeppieri \<emazep@&#8203;gmail\.com> wrote&#8203;:
> On Sat\, Nov 16\, 2013 at 6&#8203;:46 AM\, Ronald J Kimball \<rjk@&#8203;tamias\.net> wrote&#8203;:
>> On Sat\, Nov 16\, 2013 at 12&#8203;:21&#8203;:18AM \+0100\, Emanuele Zeppieri wrote&#8203;:
>>>
>>> >>    if \( my @&#8203;CAPTURE = /$\_\[0\]/ \) \{ f\( defined $1 ? @&#8203;CAPTURE : \(\) \) \}
>>> >
>>> > Yes\, that works\, or perhaps
>>> >
>>> >    if \(my @&#8203;CAPTURE = /$re/\) \{
>>> >        @&#8203;CAPTURE = \(\) if not defined $1;
>>> >        do\_something\(@&#8203;CAPTURE\);
>>> >    \}
>>> >
>>> > This does seem cumbersome\, though\, and involves looking at both $1 and
>>> > the list of captures returned by the match\.
>>> >
>>
>> You should be aware that you cannot depend on any particular capture being
>> defined\, even if the match is successful&#8203;:
>>
>> $\_ = "b";
>> my $re = qr/\(a\)|\(b\)/;
>>
>> if \(my @&#8203;CAPTURE = /$re/\) \{
>>   foreach my $i \(0 \.\. $\#CAPTURE\) \{
>>     print '$'\, $i \+ 1\, "&#8203;: "\,
>>           defined $CAPTURE\[$i\] ? "" : "not "\, "defined\\n";
>>   \}
>> \}
>>
>> Ronald
>
> The error was mine of course\.
>
> However it's just a matter of finding a reliable indicator that a
> capture has happened\, and the idea still stands&#8203;:
>
> if \( my @&#8203;CAPTURE = /$re/ \) \{
>     do\_something\( $\#\- ? @&#8203;CAPTURE : \(\) \)
> \}

Or rather&#8203;:

if \( my @&#8203;CAPTURE = /$re/ \) \{
&nbsp;    do\_something\( $\#\+ ? @&#8203;CAPTURE : \(\) \)
\}

otherwise the case of a successful match but no matching subgroups
\(only undef elements in @&#8203;CAPTURE\) is missing\.

Sorry for the noise\!

**Comment by p5pRT** [comment]
### From @rjbs

\* Kent Fredric \<kentfredric@&#8203;gmail\.com> \[2013\-11\-16T16&#8203;:16&#8203;:46\]
> On 17 November 2013 09&#8203;:53\, Father Chrysostomos \<sprout@&#8203;cpan\.org> wrote&#8203;:
> > Why match?  Why not just qr\(\.\.\.\)\->\($str\)?
> >
> > \(Then we can forget smart\-match altogether\, since \->\(\.\.\.\) is our match
> > operator\.\)
> 
> 
> Maybe because it lends itself to later enhancements\, ie&#8203;:
> 
> qr\(\.\.\.\)\->replace\(sub\{

qr\(\.\.\)\->\(\.\.\.\) does not rule out qr\(\.\.\.\)\->replace\(\.\.\.\)

\-\- 
rjbs

**Comment by p5pRT** [comment]
### From @demerphq

On 22 November 2013 04&#8203;:21\, Ricardo Signes \<perl\.p5p@&#8203;rjbs\.manxome\.org> wrote&#8203;:
> \* Kent Fredric \<kentfredric@&#8203;gmail\.com> \[2013\-11\-16T16&#8203;:16&#8203;:46\]
>> On 17 November 2013 09&#8203;:53\, Father Chrysostomos \<sprout@&#8203;cpan\.org> wrote&#8203;:
>> > Why match?  Why not just qr\(\.\.\.\)\->\($str\)?
>> >
>> > \(Then we can forget smart\-match altogether\, since \->\(\.\.\.\) is our match
>> > operator\.\)
>>
>>
>> Maybe because it lends itself to later enhancements\, ie&#8203;:
>>
>> qr\(\.\.\.\)\->replace\(sub\{
>
> qr\(\.\.\)\->\(\.\.\.\) does not rule out qr\(\.\.\.\)\->replace\(\.\.\.\)

The rules for parsing the right side of a regex are special cased\.

I fear adding support for the same special cases in a method call is
not going to fly\.

s/\(f\)o\(o\)/$1$2/

how would you write that with method call notation without requiring
us to special case parse the method call?

qr/\(f\)o\(o\)/\->replace\("$1$2"\);

isn't going to work out of the box at all\.

Yves


\-\- 
perl \-Mre=debug \-e "/just|another|perl|hacker/"

**Comment by p5pRT** [comment]
### From @kentfredric

On 22 November 2013 20&#8203;:55\, demerphq \<demerphq@&#8203;gmail\.com> wrote&#8203;:
> qr/\(f\)o\(o\)/\->replace\("$1$2"\);
>
> isn't going to work out of the box at all\.


That would of course not work\. It would have to be implemented in terms of

qr/\(f\)o\(o\)/\->replace\(sub \{
&nbsp;    "$1$2"
\}\);

Where internally\, it did :

&nbsp;   s\{\(f\)o\(o\)\}\{ $code\->\(\) \}e

At least\, thats what I'd expect to be required given the current state
of how regexp works\.

Now maybe

qr/\(f\)o\(o\)/\->replace\('$1$2'\);

Could be made work\, just the surprisingness of that seems dangerous\,
and I'd avoid it just in case\, because accidentally using "" instead
of '' and getting radically different behaviours would be bad\. \(
Especially so if $1 happened to contain the string literal '$2'  \)



\-\- 
Kent

**Comment by p5pRT** [comment]
### From @kentfredric

On 22 November 2013 23&#8203;:15\, Kent Fredric \<kentfredric@&#8203;gmail\.com> wrote&#8203;:
> qr/\(f\)o\(o\)/\->replace\(sub \{
>     "$1$2"
> \}\);


Actually\, on second thought\, even that would not always be what you
wanted\, and you might instead want :

&nbsp;  qr/\(f\)o\(o\)/\->replace\(sub \{
&nbsp;     my $match = shift;
&nbsp;     return \#somethingelsehere
&nbsp;  \}\);

And that could lend itself to doing something like&#8203;:

&nbsp;   $match\->iterpolate\("%1helloworld%2"\);

With the communicative property&#8203;:

if\( my $result = qr/\(f\)o\(o\)/\->match\($string\) \) \{
&nbsp;      print $result\->interpolate\("%1helloworld%2"\);
\}

Essentially\, making a sane way to pass state around from match results
without worrying about variable clobbering\.

I don't really know\, I must be tired or something\.
\-\- 
Kent

**Comment by p5pRT** [comment]
### From @demerphq

On 22 November 2013 11&#8203;:24\, Kent Fredric \<kentfredric@&#8203;gmail\.com> wrote&#8203;:
> On 22 November 2013 23&#8203;:15\, Kent Fredric \<kentfredric@&#8203;gmail\.com> wrote&#8203;:
>> qr/\(f\)o\(o\)/\->replace\(sub \{
>>     "$1$2"
>> \}\);
>
>
> Actually\, on second thought\, even that would not always be what you
> wanted\, and you might instead want :
>
>   qr/\(f\)o\(o\)/\->replace\(sub \{
>      my $match = shift;
>      return \#somethingelsehere
>   \}\);
>
> And that could lend itself to doing something like&#8203;:
>
>    $match\->iterpolate\("%1helloworld%2"\);
>
> With the communicative property&#8203;:
>
> if\( my $result = qr/\(f\)o\(o\)/\->match\($string\) \) \{
>       print $result\->interpolate\("%1helloworld%2"\);
> \}
>
> Essentially\, making a sane way to pass state around from match results
> without worrying about variable clobbering\.

Exactly\. The delayed interpolation of the rhs of a s/// is hard to fit
into the normal model of perl code and calling conventions\.

The only way to do so is require people pass in a closure\. IMO hardly
an improvement\.

Yves



\-\- 
perl \-Mre=debug \-e "/just|another|perl|hacker/"

**Comment by p5pRT** [comment]
### From @epa

I believe this has now been implemented as @&#8203;\{^CAPTURE\} in perl 5\.26 \(thanks demerphq\!\) so the bug can be closed\.

**Comment by p5pRT** [comment]
### From @jkeenan

On Fri\, 09 Jun 2017 15&#8203;:46&#8203;:33 GMT\, ed wrote&#8203;:
> I believe this has now been implemented as @&#8203;\{^CAPTURE\} in perl 5\.26
> \(thanks demerphq\!\) so the bug can be closed\.

Confirmed\.

\#\#\#\#\#
$ perl \-v | head \-2 | tail \-1
This is perl 5\, version 26\, subversion 0 \(v5\.26\.0\) built for x86\_64\-linux

$ cat 120521\-capture\.pl 
\# perl
use warnings;
use 5\.26\.0;

$\_ = 'abc';
/\(\.\)\(\.\)\(\.\)/ or die;
say join\(' ' => \($1\,$2\,$3\)\);

$\_ = 'def';
/\(\.\)\(\.\)\(\.\)/ or die;
say "@&#8203;\{^CAPTURE\}";

$ perl 120521\-capture\.pl 
a b c
d e f
\#\#\#\#\#

Marking ticket Resolved\.

\-\- 
James E Keenan \(jkeenan@&#8203;cpan\.org\)

**Comment by p5pRT** [comment]
@jkeenan - Status changed from 'open' to 'resolved'
